server/master/src/main/java/org/apache/accumulo/master/upgrade/Upgrader9to10.java

```diff
@@ -500,4 +506,127 @@ public void upgradeDirColumns(ServerContext ctx, Ample.DataLevel level) {
+      FileSystem fs = FileSystem.get(hadoopConf);
+
+      scanner.fetchColumnFamily(MetadataSchema.TabletsSection.DataFileColumnFamily.NAME);
+      for (Entry<Key,Value> entry : scanner) {
+        Key key = entry.getKey();
+        Text filePath = key.getColumnQualifier();
+        if (!filePath.toString().contains(":")) {
+          // found relative paths so get the property used to build the absolute paths
+          if (volumeProp == null || volumeProp.isBlank()) {
+            throw new IllegalArgumentException(
+                "Missing required property " + Property.INSTANCE_VOLUMES_UPGRADE_RELATIVE.getKey());
+          }
+          String metaEntry = key.getColumnQualifier().toString();
+          Path relPath = resolveRelativePath(metaEntry, key);
+          Path absPath = new Path(volumeProp, relPath);
+          if (fs.exists(absPath)) {
+            log.debug("Changing Tablet File path from {} to {}", metaEntry, absPath);
+            Mutation m = new Mutation(key.getRow());
+            // add the new path
+            m.at().family(key.getColumnFamily()).qualifier(absPath.toString())
+                .visibility(key.getColumnVisibility()).timestamp(key.getTimestamp())
+                .put(entry.getValue());
+            // delete the old path
+            m.at().family(key.getColumnFamily()).qualifier(key.getColumnQualifierData().toArray())
+                .visibility(key.getColumnVisibility()).timestamp(key.getTimestamp()).delete();
+            writer.addMutation(m);
+          } else {
+            throw new IllegalArgumentException(
+                "Relative Tablet file " + relPath + " not found at " + absPath);
+          }
+        }
+      }
+    } catch (MutationsRejectedException | TableNotFoundException e) {
+      throw new RuntimeException(e);
+    } catch (IOException ioe) {
+      throw new UncheckedIOException(ioe);
+    }
+  }
+
+  /**
+   * Check if table has any relative paths, return false if none are found. When a relative path is
+   * found, check existence of the file path constructed from the upgrade property + relative path
+   */
+  public static boolean checkForRelativePaths(AccumuloClient client, Configuration hadoopConf,
+      String tableName, String upgradeProperty) {
+    boolean hasRelatives = false;
+
+    try (Scanner scanner = client.createScanner(tableName, Authorizations.EMPTY)) {
+      FileSystem fs = FileSystem.get(hadoopConf);
+      log.info("Looking for relative paths in {}", tableName);
+      scanner.fetchColumnFamily(MetadataSchema.TabletsSection.DataFileColumnFamily.NAME);
+      for (Entry<Key,Value> entry : scanner) {
+        Key key = entry.getKey();
+        Text filePath = key.getColumnQualifier();
+        if (!filePath.toString().contains(":")) {
+          // found relative paths so verify the property used to build the absolute paths
+          hasRelatives = true;
+          if (upgradeProperty == null || upgradeProperty.isBlank()) {
```

Comment:
Will the upgradeProperty be normalized by the `new Path` code? If not, it might be useful to error out on things like having multiple `//` (except if it immediately follows the first `:`, as in `hdfs://`) or having any `../` present in the string. If `Path` normalizes this, then this is less important.

Reply:
I believe so... after this check, I create the path: ```Path absPath = new Path(upgradeProperty, relPath);```

Reply:
That might just do a simple concatenation with the appropriate file separator. It might not do normalization.

Reply:
Looking at the source, there is some normalization done by [Path](https://github.com/apache/hadoop/blob/trunk/hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/fs/Path.java#L183) using the java URI object [here](https://github.com/apache/hadoop/blob/trunk/hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/fs/Path.java#L260).  And as far as I can tell that source is the same as the source I downloaded for 3.1.2: http://archive.apache.org/dist/hadoop/common/hadoop-3.1.2/hadoop-3.1.2-src.tar.gz

Reply:
I also confirmed that extra intermediate ```/``` get removed by this normalization, with my last update adding an extra erroneously to one of the relative paths:
```tables//1b/default_tablet/A000001c.rf``` becomes ```tables/1b/default_tablet/A000001c.rf```

It won't hurt to add some erroneous paths to the unit test though, now that I have one.

Reply:
> It won't hurt to add some erroneous paths to the unit test though, now that I have one.

Agreed.